clock/clock_mutex.go

```diff
@@ -0,0 +1,155 @@
+	defer rwMutex.Unlock()
+	provider = &frozenTime{now: now}
+	return Unfreezer{}
+}
+
+type Unfreezer struct{}
+
+func (u Unfreezer) Unfreeze() {
+	Unfreeze()
+}
+
+// Unfreeze reverses effect of Freeze.
+func Unfreeze() {
+	rwMutex.Lock()
+	defer rwMutex.Unlock()
+	provider = realtime
+}
+
+// Realtime returns a clock provider wrapping the SDK's time package. It is
+// supposed to be used in tests when time is frozen to schedule test timeouts.
+func Realtime() Clock {
+	return realtime
+}
+
+// Makes the deterministic time move forward by the specified duration, firing
+// timers along the way in the natural order. It returns how much time has
+// passed since it was frozen. So you can assert on the return value in tests
+// to make it explicit where you stand on the deterministic time scale.
+func Advance(d time.Duration) time.Duration {
+	rwMutex.RLock()
+	ft, ok := provider.(*frozenTime)
+	rwMutex.RUnlock()
+	if !ok {
+		panic("Freeze time first!")
+	}
+	ft.advance(d)
+	return Now().UTC().Sub(frozenAt)
+}
+
+// Wait4Scheduled blocks until either there are n or more scheduled events, or
+// the timeout elapses. It returns true if the wait condition has been met
+// before the timeout expired, false otherwise.
+func Wait4Scheduled(count int, timeout time.Duration) bool {
+	rwMutex.RLock()
+	defer rwMutex.RUnlock()
+	return provider.Wait4Scheduled(count, timeout)
+}
+
+// Now see time.Now.
+func Now() time.Time {
+	rwMutex.RLock()
+	defer rwMutex.RUnlock()
+	return provider.Now()
+}
+
+// Sleep see time.Sleep.
+func Sleep(d time.Duration) {
+	rwMutex.RLock()
+	defer rwMutex.RUnlock()
+	provider.Sleep(d)
+}
+
+// After see time.After.
+func After(d time.Duration) <-chan time.Time {
+	rwMutex.RLock()
+	defer rwMutex.RUnlock()
+	return provider.After(d)
+}
+
+// Timer see time.Timer.
+type Timer interface {
+	C() <-chan time.Time
+	Stop() bool
+	Reset(d time.Duration) bool
+}
+
+// NewTimer see time.NewTimer.
+func NewTimer(d time.Duration) Timer {
+	return provider.NewTimer(d)
+}
+
+// AfterFunc see time.AfterFunc.
+func AfterFunc(d time.Duration, f func()) Timer {
+	return provider.AfterFunc(d, f)
+}
+
+// Ticker see time.Ticker.
+type Ticker interface {
+	C() <-chan time.Time
+	Stop()
+}
+
+// NewTicker see time.Ticker.
+func NewTicker(d time.Duration) Ticker {
+	return provider.NewTicker(d)
+}
+
+// Tick see time.Tick.
+func Tick(d time.Duration) <-chan time.Time {
+	return provider.Tick(d)
+}
+
+// NewStoppedTimer returns a stopped timer. Call Reset to get it ticking.
+func NewStoppedTimer() Timer {
+	t := NewTimer(42 * time.Hour)
+	t.Stop()
+	return t
+}
+
+// Clock is an interface that mimics the one of the SDK time package.
+type Clock interface {
+	Now() time.Time
+	Sleep(d time.Duration)
+	After(d time.Duration) <-chan time.Time
+	NewTimer(d time.Duration) Timer
+	AfterFunc(d time.Duration, f func()) Timer
+	NewTicker(d time.Duration) Ticker
+	Tick(d time.Duration) <-chan time.Time
+	Wait4Scheduled(n int, timeout time.Duration) bool
+}
```

Comment:
@vtopc Yea that makes. I moved that code into a `base.go` file without a build tag.

Reply:
Okay, I think I may have misunderstood since the code you were referencing was deleted. I think we just came up with a better way of doing this, and @Takumi2008 will push that shortly.

Reply:
I moved the interface definitions to `interface.go` and removed `base.go`.  So we have the interface definitions in one file and the 2 different implementations in their own files.